### DISCUSSION TITLE
Modify user based on authentication method

### DISCUSSION THREAD

**Discussion by nfreader** [symfony/symfony#49875] [Q&A] [answered]
Hello! 

I'm not sure what combination of words to search for to answer this, so I'm hoping someone here can help.

Right now, users can log into my application in two ways: username and password (from the [default security bundle](https://symfony.com/doc/current/security.html#the-user)), or with a security key (via the [webauthn](https://github.com/web-auth/webauthn-framework) [bundle](https://webauthn-doc.spomky-labs.com/symfony-bundle/the-symfony-way)). Ideally, I'd be able to dynamically add a role to the user, based on how they authenticated to the application. Any hints as to where I should be looking to accomplish this?

Thanks! :)

**Comment by ghost** [comment]
Never done that, but I think that you can write a custom authenticator and configure this for a given firewalls. Then add a role dynamically inside `\Symfony\Component\Security\Guard\AuthenticatorInterface::getUser()` (in old symfony security system, in new probably something similar but I haven't used new system so far). Yo can extend/decorate currently using authenticator not to duplicate the code, just to add this dynamic roles. 
Or maybe it would be more correct to decorate currently used UserProvider (dynamic roles as constructor arguments), then create two services - one instance per firewall.

P.S. if your user model is a persistent model, e.g. doctrine entity you have to be careful - don't add dynamic roles to the array field that is mapped to database but to a separate field and from `getRoles()` return a merged array rolesInDatabase+dynamicRoles.

**Comment by wouterj** [comment, marked as answer]
I recommend not using roles for this. A "role" should represent the role of that user in the application (e.g. is it a user, an adminstrator, etc.). Using it for another purpose can break in unpredictable ways (e.g. adding a new role will end all active sessions and require the user to authenticate again).

You can however differentiate based on information in the security token. E.g. the webauthn authenticator creates a [`WebauthnToken`](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/symfony/src/Security/Authentication/Token/WebauthnToken.php) that is stored in the session, whereas the build-in form login uses a [`UsernamePasswordToken`](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Security/Core/Authentication/Token/UsernamePasswordToken.php).

You can write a [custom voter](https://symfony.com/doc/current/security/voters.html#creating-the-custom-voter) that checks whether `TokenStorageInterface::getToken()` returns an instance of `WebauthnToken` and use it to check permissions (e.g. `isGranted('IS_AUTHENTICATED_WEBAUTHN')`).

**Reply by nfreader** [reply]
Ah, thank you very much. I skimmed the voter documentation but I didn't realize it would let me use `isGranted` like that (which was the genesis of my question in the first place).
